### PR TITLE
Virtual memory allocators, Allocator fixes, Spinlock

### PR DESCRIPTION
**Pull request by jkunstwald**
- Add `virtual_linear_allocator`, `virtual_stack_allocator`
  - Virtual memory allocators (linux and win32 implementations)
- Add `cc::spin_lock`, `cc::lock_guard` (ported from task dispatcher)
- Delete `cc::scratch_allocator` due to further undiagnosed bugs
- Fix realloc behavior of linear and stack allocs
- Add synced versions of virtual allocs, make lock a template parameter of `synced_` alloc versions

### CODE REVIEW

File: src/clean-core/allocator.hh [hunk 1]
```diff
@@ -142,55 +165,126 @@ struct stack_allocator final : allocator
...
 };
 
-
-/// for short lived allocations serviced from a ring buffer
-/// can optionally fall back to an external allocator if overcommited
-struct scratch_allocator final : allocator
+/// linear allocator operating in virtual memory
+/// reserves pages on init, commits pages on demand
+/// only frees pages if explicitly called
+struct virtual_linear_allocator final : allocator
```
Comment (Philip-Trettner) [inline src/clean-core/allocator.hh]: Hm we now have a lot of "niche" allocators in this general allocator header that is included everywhere. I think we should all/most specialized allocators into a separate file, probably even experimental for now

File: src/clean-core/allocator.hh [hunk 2]
```diff
@@ -142,55 +165,126 @@ struct stack_allocator final : allocator
...
+    explicit stack_allocator(span<std::byte> buffer) : _buffer_begin(buffer.data()), _head(buffer.data()), _buffer_end(buffer.data() + buffer.size())
+    {
+    }
 
 private:
     std::byte* _buffer_begin = nullptr;
     std::byte* _head = nullptr;
     std::byte* _buffer_end = nullptr;
-    std::byte* _last_alloc = nullptr;
+    int32_t _last_alloc_id = 0;
```
Comment (Philip-Trettner) [inline src/clean-core/allocator.hh]: notice how this leaves a 32bit padding hole

File: src/clean-core/native/memory.hh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// frees a virtual memory range, the result of reserve_virtual_memory
+// also decommits all physical regions inside
+void free_virtual_memory(std::byte* ptr, size_t size_bytes);
+
+// commits a region of pages inside a reserved, virtual memory range
+void commit_physical_memory(std::byte* ptr, size_t size_bytes);
+
+// decommits a region of pages inside a reserved, virtual memory range
+void decommit_physical_memory(std::byte* ptr, size_t size_bytes);
+}
```
Comment (Philip-Trettner) [inline src/clean-core/native/memory.hh:57]: might miss a format

File: src/clean-core/native/memory.hh [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <cstddef>
+
+namespace cc
+{
+// reserves a range of pages in virtual memory
+std::byte* reserve_virtual_memory(size_t size_bytes);
```
Comment (Philip-Trettner) [inline src/clean-core/native/memory.hh:11]: how would my use case of "I want 1 TB memory that is commit-on-use" work?

File: src/clean-core/native/memory.hh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#pragma once
```
Comment (Philip-Trettner) [inline src/clean-core/native/memory.hh:1]: I'd like to discuss a owning RAII version of these and if we can a bit of impedance to the "raw" functions (as they are so easy to use wrongly. maybe move in a sub namespace or something).

File: src/clean-core/spin_lock.hh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    spin_lock& operator=(spin_lock const& other) = delete;
+    spin_lock& operator=(spin_lock&& other) noexcept = delete;
+
+private:
+    std::atomic_bool _is_locked = {false};
+};
+
+// simple replacement for std::lock_guard
+template <typename T>
+struct [[nodiscard]] lock_guard
```
Comment (Philip-Trettner) [inline src/clean-core/spin_lock.hh]: it's a bit weird to have this in spin_lock.hh

File: src/clean-core/spin_lock.hh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+{
+    CC_FORCE_INLINE explicit lock_guard(T& mutex) : _mutex(mutex) { _mutex.lock(); }
+    CC_FORCE_INLINE ~lock_guard() noexcept { _mutex.unlock(); }
```
Comment (Philip-Trettner) [inline src/clean-core/spin_lock.hh]: dtors are noexcept by default

File: src/clean-core/allocator.hh (hunk 1 above)
Comment (jkunstwald) [reply src/clean-core/allocator.hh]: Yes, in fact an allocators/ folder might be a good idea with separate headers per type. We could still have "sink headers" for multiple ones

File: src/clean-core/allocator.hh (hunk 2 above)
Comment (jkunstwald) [reply src/clean-core/allocator.hh]: This was a bugfix not an optimization, the id is stored in alloc headers

File: src/clean-core/native/memory.hh (hunk 5 above)
Comment (jkunstwald) [reply src/clean-core/native/memory.hh:1]: These functions are used by the vmem allocators, so convenient usage like that should just be done via them and existing container types. I was thinking about a namespace as well but this is already in the native/ folder

File: src/clean-core/native/memory.hh (hunk 4 above)
Comment (jkunstwald) [reply src/clean-core/native/memory.hh:11]: On linux that's just malloc(1TB) iirc, these are for explicit commit/decommit as is done by the vmem allocs

File: src/clean-core/spin_lock.hh (hunk 6 above)
Comment (jkunstwald) [reply src/clean-core/spin_lock.hh]: Yeah, might deserve a separate header

File: src/clean-core/detail/vector_base.hh [hunk 8]
```diff
@@ -166,24 +167,56 @@ public:
...
+            if constexpr (collection_traits<Range>::has_size)
+                this->reserve(_size + cc::collection_size(range));
+
             for (auto&& v : range)
                 this->push_back(v);
         }
     }
 
+    /// adds an element at the given index, moves
+    void insert_at_index(T const& value, size_t index) { insert_range_at_index(cc::span<T const>(value), index); }
```
Comment (Philip-Trettner) [inline src/clean-core/detail/vector_base.hh:199]: maybe `insert_at`  to be consistent with `remove_at` ?

**Pull request merged**